### PR TITLE
[usb_testutils] Streaming changes for suspend-resume

### PR DESCRIPTION
**Pull request by alees24**
Commit 1: Move transmission state into a separate structure, in preparation for suspend-resume changes; no functional change. Diff tools do not render this well.

Commit 2: Introduce a concept of 'committed' transmission state, so that we may rewind reliable transfer streams properly to the point at which Suspending occurred, even though packet generation has run on ahead. This is required for Bulk and Interrupt transfer streams to resume at the point expected by the host-side streaming code.

Updated usbdev_stream/mixed/iso_test to use modified API and to correct the interface counts in the configuration descriptors. All may now be run manually on the CW310 and the latest host-side streaming code (yet to be presented) reads the test descriptor from the device code and adapts accordingly.

### CODE REVIEW

**Review by marnovandermaas — COMMENTED** [review]
This code looks reasonable to me, thanks for splitting out the "non-functional" commit. A couple of questions:

- What are the instructions to build this code?
- What are the instructions to test this code?
- Would it be useful for someone from the software team to have a read through the code?

**Comment by alees24** [comment]
> This code looks reasonable to me, thanks for splitting out the "non-functional" commit. A couple of questions:
> 
>     * What are the instructions to build this code?

It is built into the existing 3 streaming tests, including the top-level regressions, eg.
```
bazel test --test_output=streamed //sw/device/tests:usbdev_stream_test[_fpga_cw310_test_rom]
```
will compile and build the code...

>     * What are the instructions to test this code?

... however, the suspend/resume functionality is not exercised in any of the already-submitted tests. The in-progress draft of the device-side suspend/resume code lives at [#18564](https://github.com/lowRISC/opentitan/pull/18564) but it requires the host-side code and manual supervision at present.

>     * Would it be useful for someone from the software team to have a read through the code?

I'm hoping that Alex will provide most of the input re USB behavior, code style etc.

**Review by marnovandermaas — APPROVED** [review]
I've tested this PR with the following commands:
```
bazel build --test_output=streamed //sw/device/tests:usbdev_stream_test
bazel build --test_output=streamed //sw/device/tests:usbdev_stream_test_fpga_cw310_test_rom
```

Both seem to complete successfully (the first one requires Vivado). I think it is fine to not be able to explicitly test the code in this PR because you have tested it manually and it is an incremental PR.

We should wait for another reviewer to review this PR before merging.

**Comment by a-will** [comment]
> I'm hoping that Alex will provide most of the input re USB behavior, code style etc.

Sorry this review has been slow in coming! I'll have a closer look tomorrow.

**Review by a-will — APPROVED** [review]
LGTM

File: sw/device/lib/testing/usb_testutils_streams.h
Comment (alees24) [reply sw/device/lib/testing/usb_testutils_streams.h]: Well, I'm not in the business of confusing or misleading people (that's just a sideline hobby), so I've renamed it to ..stream_cb_ctx_t, which I hope is clear enough.

**Pull request merged**
